File: ws/src/cpp_example/src/simple_services_client.cpp
#include <rclcpp/rclcpp.hpp>
#include <memory>
#include <robot_msgs/srv/add_two_ints.hpp>
#include <chrono>

using namespace std::chrono_literals;
using std::placeholders::_1;

class SimpleServiceClient : public rclcpp:: Node{
    public:
        SimpleServiceClient(int a, int b) : Node("simple_service_client")
        {
            client_ = create_client<robot_msgs::srv::AddTwoInts>("add_two_ints");
            auto request_ = std::make_shared<robot_msgs::srv::AddTwoInts::Request>();
            request_->a = a;
            request_->b = b;

            while(!client_->wait_for_service(1s)){
                RCLCPP_ERROR(this->get_logger(), "Service not available, waiting more thime ...");
                if(!rclcpp::ok()){
                    RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for service");
                    return ;
                }
            }
            auto result_ = client_->async_send_request(request_, std::bind(&SimpleServiceClient::responseCallback, this, _1));

        }
    private:
        rclcpp::Client<robot_msgs::srv::AddTwoInts>::SharedPtr client_;
        void responseCallback(rclcpp::Client<robot_msgs::srv::AddTwoInts>::SharedFuture future){
            if(future.valid()){
                RCLCPP_INFO_STREAM(this->get_logger(), "Service Response " << future.get()->sum);
            }else{
                RCLCPP_ERROR(this->get_logger(), "Service Failure");
            }
        };
};

int main(int argc, char* argv[]){
    rclcpp::init(argc, argv);
    if(argc != 3){
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Wrong number of arguments! Usage: simple_service client A B");
        return 1;
    }
    auto node = std::make_shared<SimpleServiceClient>(atoi(argv[1]), atoi(argv[2]));
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ws/src/py_example/py_example/simple_services_client.py
import rclpy
from rclpy.node import Node
from robot_msgs.srv import AddTwoInts
from rclpy.task import Future
import sys

class SimpleServiceClient(Node):
    def __init__(self, a:int, b:int):
        super().__init__(node_name="simple_service_client")

        self.client_ = self.create_client(AddTwoInts, "add_two_ints")
        while not self.client_.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("Service not available, waiting more...")
        
        self.req_: AddTwoInts.Request = AddTwoInts.Request()
        self.req_.a = a
        self.req_.b = b
        self.future_: Future = self.client_.call_async(self.req_)
        self.future_.add_done_callback(self.responseCallback)

    
    def responseCallback(self, future: Future):
        self.get_logger().info(f"Service Response {future.result().sum}")

def main():
    rclpy.init()
    if len(sys.argv)!= 3:
        print("Wrong number of arguments! Usage: simple_services_client A B")
        return -1
    simple_service_client = SimpleServiceClient(int(sys.argv[1]), int(sys.argv[2]))
    rclpy.spin(simple_service_client)
    simple_service_client.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()